### PR TITLE
[FEATURE] Snap geometries algorithm

### PR DESCRIPTION
**Pull request by wonder-sk**
Makes sure that any two vertices of the vector layer are at least at distance given by the threshold value. The algorithm moves nearby vertices to one location and adds vertices to segments that are passing around other vertices within the threshold. It does not remove any vertices. Also, it does not modify geometries unless needed (it does not snap coordinates to a grid).

This algorithm comes handy when doing vector overlay operations such as intersection, union or difference to prevent possible topological errors caused by numerical errors if coordinates are very close to each other.

After running the algorithm some previously valid geometries may become invalid and therefore it may be useful to run Fix geometries algorithm afterwards.

Credits for this algorithm go to the GRASS team - this is essentially Vect_snap_lines_list() ported over to QGIS code with few fixes.

# Q & A

## When would I want to use this algorithm?

When your input data may have some coordinates too close to each other and you are about to run a vector overlay algorithm. I had a layer where all geometries were valid, yet the Difference algorithm would end up with topological errors. For example, think of a geometry like this:
```
+-------+
|       |
|    +--+ A
|    |
|    +--+ B
|       |
+-------+
```
where the distance between points A and B is really small (e.g. 1e-10). Cases like this can make GEOS raise topological errors due to numerical errors. Snapping vertices to each other with a small tolerance solves the problems with minimal impact.

In general it is good to employ snapping algorithm when cleaning input data coming from unknown sources: first do the snapping, then run fix geometries algorithm to handle various other issues with geometries (including newly introduced errors due to snapping).

## Wait... We already have a snapping algorithm!

True, let me explain.

There are `QgsGeometrySnapper` and `QgsInternalGeometrySnapper` classes, both of them are wrapped by "Snap geometries to layer" algorithm in the Processing toolbox. There are two ways how the snapping can be done:
- snap geometries of layer A to geometries from reference layer B
- snap geometries of layer A to themselves

The newly added algorithm in this PR intends to replace the second variant (snapping of geometries to the layer itself).

Trying to run the original snapping algorithm with a reasonably sized data (~4 MB shapefile) first took several minutes before crashing(!) QGIS without getting any result. Running the algorithm with a small subset finished, but it did not fix my topological issues. To summarize the problems with the existing algorithm:
- it is unable to snap vertices to vertices of the same geometry - e.g. the U-shaped polygon in the example above would stay untouched
- for every single geometry it iterates over the source layer and builds the snapping index - this is making the algorithm very slow
- the snapping index data structure is quite sub-optimal and its memory requirements depend on tolerance, making it slow and prone to exhausting all memory

Considering these issues, I have realized that it is simply not possible to just fix the existing snapping algorithm. After some research I have decided to port GRASS snapping algorithm because it does not have any of the issues above. With the same sample data (~4 MB shapefile) the snapping is done within few seconds and it fixed the topological errors I was getting earlier.

## What to do with the existing snapping algorithm(s)?

The geometry snapper classes are exposed in the API and as Processing algorithms, so we will need to live with them at least until QGIS 4.

Also, the new algorithm does not handle the second variant of snapping: snapping layer A to reference layer B. It would be useful to port also this variant from GRASS - for that there is Vect_snap_line() function.

### CODE REVIEW

File: src/analysis/processing/qgsalgorithmsnapgeometries.cpp [hunk 1]
```diff
@@ -0,0 +1,411 @@
...
+
+    pnts[point].anchor = -2; // make it anchor
+    nanchors++;
+
+    // Find points in threshold
+    double x = pnts[point].x, y = pnts[point].y;
+    QgsRectangle rect( x - thresh, y - thresh, x + thresh, y + thresh );
+
+    const QList<QgsFeatureId> ids = index.intersects( rect );
+    for ( QgsFeatureId pointb : ids )
```
Comment (pblottiere) [inline src/analysis/processing/qgsalgorithmsnapgeometries.cpp]: Maybe `for ( const QgsFeatureId &pointb : ids ){...}`?

File: src/analysis/processing/qgsalgorithmsnapgeometries.cpp [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+
+    QgsRectangle rect( xmin - thresh, ymin - thresh, xmax + thresh, ymax + thresh );
+
+    // Find points
+    const QList<QgsFeatureId> fids = index.intersects( rect );
+
+    QVector<AnchorAlongSegment> newVerticesAlongSegment;
+
+    // Snap to anchor in threshold different from end points
+    for ( QgsFeatureId fid : fids )
```
Comment (pblottiere) [inline src/analysis/processing/qgsalgorithmsnapgeometries.cpp]: `for ( const QgsFeatureId &fid : fids )`?

File: src/analysis/processing/qgsalgorithmsnapgeometries.cpp [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+    {
+      int spoint = fid;
```
Comment (pblottiere) [inline src/analysis/processing/qgsalgorithmsnapgeometries.cpp]: Why using such an intermediate variable `spoint` instead of `fid`?

**Comment by nyalldawson** [comment]
@wonder-sk I haven't dug into the code yet, but I've got a general reservation with the approach taken here.

What I'm concerned about is adding a new algorithm which partially overlaps an existing algorithm. Because with this PR, we've got the existing (slow, fragile) algorithm which handles snapping but which offers all the "behavior" choices, and a new faster, stabler algorithm but which doesn't expose these options.

This isn't a great situation, because:
- Users won't know which to use. They're probably just as likely to use the existing one because it exposes more options.
- The new algorithm becomes part of stable api, and we can't remove it, even if the new code is used to replace the guts of the existing algorithm.
- People will start using this algorithm in models, and the only way we can ever remove the duplication is to take the drastic approach we did for 3.0 and declare that 3.0 models aren't compatible with 4.0. I'd really like to avoid this -- it's not nice for users to force them to recreate models.

So I'd see two options here:

1. Add the missing behavior choices to the new code, and then make the existing algorithm use the new snapping code instead of the current code.

2. Just add this as a python class for now and don't expose at all to processing until 1 can be done.

**Comment by wonder-sk** [comment]
@nyalldawson I understand your concerns, but... :-)

For InaSAFE when I have run Union algorithm between two real-world datasets, the Union algorithm (resp. the Difference sub-algorithm) was failing inside GEOS due to topology errors. There is currently no way to fix that in QGIS: the existing geometry snapper chokes on the data and crashes (and when tested with just one offending polygon, it does not even fix the problem(!)). One could use snap to grid algorithm, but that modifies _all_ the input vertices and thus potentially introduces new topological issues. I have figured out that GRASS has a robust geometry snapping algorithm that was both fast and it fixed the input dataset (only the bits that needed fixing - it keeps the rest of data untouched).

> we've got the existing (slow, fragile) algorithm which handles snapping but which offers all the "behavior" choices, and a new faster, stabler algorithm but which doesn't expose these options.

I would argue that the extra options are not really needed. GRASS does not have any such options and it is doing just fine :-) Whenever I look at the list of the extra options how to snap features in the existing algorithm, I do not know what are they doing and I need 30 minutes of looking into code to understand the differences.

The extra options offered by the existing geometry snapping alg cannot be incorporated into the algorithm from GRASS without a total rewrite.

> Users won't know which to use.

They will try one and if it works, for them they will stick to it. If not, they will try the other one. This is already the case of similar/same algorithms with different alg providers - not ideal, but people will eventually find their way.

> People will start using this algorithm in models, and the only way we can ever remove the duplication is to take the drastic approach we did for 3.0 and declare that 3.0 models aren't compatible with 4.0.

I think we will have to do it anyway - just like we use the major versions to clean up the API, I think we should also allow such cleanups for processing algorithms. Not as drastic as with 3.0 though - we should document the removed/changed algorithms properly and possibly help users migrate their models...

> 1. Add the missing behavior choices to the new code, and then make the existing algorithm use the new snapping code instead of the current code.
>
> 2. Just add this as a python class for now and don't expose at all to processing until 1 can be done.

As said above, the new code will probably never support the behaviors of the existing algorithm, so unfortunately I don't think either of these options are feasible.

To summarize: I would rather accept some overlap of functionality between the existing snapping alg and the new alg than not having a properly working snapping algorithm at all. For long time people have resorted to GRASS to fix their data, they should not need to anymore if they can have a robust snapping algorithm directly in QGIS...

**Comment by nyalldawson** [comment]
@wonder-sk 
> For InaSAFE when I have run Union algorithm between two real-world datasets, the Union algorithm (resp. the Difference sub-algorithm) was failing inside GEOS due to topology errors. There is currently no way to fix that in QGIS: the existing geometry snapper chokes on the data and crashes (and when tested with just one offending polygon, it does not even fix the problem(!)). One could use snap to grid algorithm, but that modifies all the input vertices and thus potentially introduces new topological issues. I have figured out that GRASS has a robust geometry snapping algorithm that was both fast and it fixed the input dataset (only the bits that needed fixing - it keeps the rest of data untouched).

That's fair enough, but I think this decision shouldn't be about specifics of one algorithm's implementation (we both can agree that there's issues with the current algorithm), and rather about the high-level design/consequences of overlapping algorithms.

> I would argue that the extra options are not really needed. GRASS does not have any such options and it is doing just fine :-)

I don't think that argument holds up. There's a lot which QGIS does which GRASS doesn't do.

>  Whenever I look at the list of the extra options how to snap features in the existing algorithm, I do not know what are they doing and I need 30 minutes of looking into code to understand the differences.

Strongly disagree! I added those "end point only"  options to allow the snapping algorithm to be used to clean up road networks and make them topologically consistent. The extra options make the snap algorithm a better match for ArcGIS' snap geoprocessing tool, which has similar options.


> They will try one and if it works, for them they will stick to it. If not, they will try the other one.

That's just a cause of user frustration, loss of faith in QGIS' stability, and UI/application complexity. My experience would be more like "they will try one, if it doesn't work, they'll assume QGIS just can't do this". Heck, I see this all the time with plugins (plugin X is broken, but instead of trying alternative plugin Y, a user will just stick to plugin X and get frustrated with QGIS).

> This is already the case of similar/same algorithms with different alg providers - not ideal, but people will eventually find their way.

Yes, but there's a clear separation between algorithms offered by different providers. They have different icons and are separated into different groups (and also have different naming conventions, etc). 

People will start using this algorithm in models, and the only way we can ever remove the duplication is to take the drastic approach we did for 3.0 and declare that 3.0 models aren't compatible with 4.0.

> I think we will have to do it anyway - just like we use the major versions to clean up the API, I think we should also allow such cleanups for processing algorithms. Not as drastic as with 3.0 though - we should document the removed/changed algorithms properly and possibly help users migrate their models...

Maybe, but I'd only be comfortable if this was a LAST FINAL END-of-the-WORLD resort!

>> Add the missing behavior choices to the new code, and then make the existing algorithm use the new snapping code instead of the current code.

>> Just add this as a python class for now and don't expose at all to processing until 1 can be done.

> As said above, the new code will probably never support the behaviors of the existing algorithm, so unfortunately I don't think either of these options are feasible.

Why not go with an ugly approach: unify the algorithms, and if the extra options from the existing algorithm are NOT used, then use the ported grass code. It's ugly, but I think it's better than duplicate algorithms.

> To summarize: I would rather accept some overlap of functionality between the existing snapping alg and the new alg than not having a properly working snapping algorithm at all. For long time people have resorted to GRASS to fix their data, they should not need to anymore if they can have a robust snapping algorithm directly in QGIS...

Sorry, but I respectfully disagree. We don't use this approach elsewhere ("WCS provider is not working in some situations, so let's add a second provider for WCS which handles these situations"). Rather, we fix/replace the existing broken code!

**Comment by wonder-sk** [comment]
> Why not go with an ugly approach: unify the algorithms, and if the extra options from the existing algorithm are NOT used, then use the ported grass code. It's ugly, but I think it's better than duplicate algorithms.

I was looking at that option today and it could be a good compromise...

We would however need a new mode - I was looking at the existing modes again and the way the GRASS alg works does not fit any of the existing modes. The alg here could be described as "Snap nodes to anchors, insert extra vertices (single layer only)" mode - it first takes all vertices, marks some of them as anchors, snaps non-anchor vertices to anchors and adds vertices to segments that are within tolerance of anchors. (The existing modes "prefer aligning nodes" and "prefer closest point" always consider both vertices and segments, while the new mode only considers vertices.)

Would that work for you?

**Comment by nyalldawson** [comment]
> Would that work for you?

That works for me. I'm happy for you to make that the default mode too (and just give it a friendly name like "automatic")

File: src/analysis/processing/qgsalgorithmsnapgeometries.cpp (hunk 1 above)
Comment (wonder-sk) [reply src/analysis/processing/qgsalgorithmsnapgeometries.cpp]: ```const QgsFeatureId &``` is actually worse since QgsFeatureId is a 64-bit int, so it is same length as a reference, but a reference would require extra dereferencing...

**Comment by wonder-sk** [comment]
Updated. I refrained from making it the default mode because it only works with a single input layer, so it could be confusing for the default...

**Comment by DelazJ** [comment]
Just a UI inconsistency I found today: the Union algorithm allows to run on a single layer, hence the "overlay layer" is tagged as optional. Given that the "snap geometries to layer" behaves the same way, could the "reference layer" option be set as optional , hence empty by default?

File: python/plugins/processing/algs/qgis/SnapGeometries.py [hunk 4]
```diff
@@ -119,6 +124,10 @@ def processAlgorithm(self, parameters, context, feedback):
                     sink.addFeature(f)
                 processed += 1
                 feedback.setProgress(processed * total)
+        elif mode == 7:
+            # input layer == ref layer
+            modified_count = QgsGeometrySnapperSingleSource.run(source, sink, tolerance, feedback)
+            feedback.pushInfo(self.tr('Snapped {} geometries.').format(modified_count))
```
Comment (nyalldawson) [inline python/plugins/processing/algs/qgis/SnapGeometries.py]: Ideally this would be a numeric output from the alg, but the other modes prevent that.






^^ thinking aloud only

File: src/analysis/vector/qgsgeometrysnappersinglesource.cpp [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+#include "qgsfeaturesource.h"
+#include "qgsfeedback.h"
+#include "qgsgeometrycollection.h"
+#include "qgsgeometryutils.h"
+#include "qgslinestring.h"
+#include "qgspolygon.h"
+#include "qgsspatialindex.h"
+
+//! record about vertex coordinates and index of anchor to which it is snapped
+typedef struct
```
Comment (nyalldawson) [inline src/analysis/vector/qgsgeometrysnappersinglesource.cpp]: This is a rather old-school approach to defining structs! ;)

File: src/analysis/vector/qgsgeometrysnappersinglesource.cpp [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+      std::sort( newVerticesAlongSegment.begin(), newVerticesAlongSegment.end(), []( const AnchorAlongSegment & p1, const AnchorAlongSegment & p2 )
+      {
+        return ( p1.along < p2.along ? -1 : ( p1.along > p2.along ) );
+      } );
+
+      // insert new vertices
+      for ( int i = 0; i < newVerticesAlongSegment.count(); i++ )
+      {
+        int anchor = newVerticesAlongSegment[i].anchor;
+        newPoints << QgsPoint( pnts[anchor].x, pnts[anchor].y, 0 );
```
Comment (nyalldawson) [inline src/analysis/vector/qgsgeometrysnappersinglesource.cpp]: Just for my understanding -- this algorithm will lost m/z values, right?

File: src/analysis/vector/qgsgeometrysnappersinglesource.cpp [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+
+  int count = 0;
+  int totalCount = source.featureCount() * 2;
+
+  // step 1: record all point locations in a spatial index + extra data structure to keep
+  // reference to which other point they have been snapped to (in the next phase).
+
+  QgsSpatialIndex index;
+  QVector<AnchorPoint> pnts;
+  QgsFeatureIterator fi = source.getFeatures();
```
Comment (nyalldawson) [inline src/analysis/vector/qgsgeometrysnappersinglesource.cpp]: buildSnapIndex doesn't require attributes

File: src/analysis/vector/qgsgeometrysnappersinglesource.cpp [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+  // Go through all lines and:
+  //   1) for all vertices: if not anchor snap it to its anchor
+  //   2) for all segments: snap it to all anchors in threshold (except anchors of vertices of course)
+
+  int modified = 0;
+  QgsFeature f;
+  fi = source.getFeatures();
+  while ( fi.nextFeature( f ) )
+  {
+    QgsGeometry geom = f.geometry();
```
Comment (nyalldawson) [inline src/analysis/vector/qgsgeometrysnappersinglesource.cpp]: Check for feedback cancelation

File: src/analysis/vector/qgsgeometrysnappersinglesource.cpp [hunk 9]
```diff
@@ -0,0 +1,345 @@
...
+
+
+static void buildSnapIndex( QgsFeatureIterator &fi, QgsSpatialIndex &index, QVector<AnchorPoint> &pnts, QgsFeedback *feedback, int &count, int totalCount )
+{
+  QgsFeature f;
+  int pntId = 0;
+
+  while ( fi.nextFeature( f ) )
+  {
+    QgsGeometry g = f.geometry();
```
Comment (nyalldawson) [inline src/analysis/vector/qgsgeometrysnappersinglesource.cpp]: Check for feedback cancelation

File: src/analysis/vector/qgsgeometrysnappersinglesource.cpp (hunk 5 above)
Comment (wonder-sk) [reply src/analysis/vector/qgsgeometrysnappersinglesource.cpp]: One can nearly smell the old wood :-) but if you prefer more modern style, I can change that

File: src/analysis/vector/qgsgeometrysnappersinglesource.cpp (hunk 6 above)
Comment (wonder-sk) [reply src/analysis/vector/qgsgeometrysnappersinglesource.cpp]: Not really - only the newly created points will not have m/z values - and that's fair I think (I don't think interpolating them would universally work as there are two conflicting candidates 1. from the segment vs 2. from the anchor)

**Comment by wonder-sk** [comment]
> Just a UI inconsistency I found today: the Union algorithm allows to run on a single layer, hence the "overlay layer" is tagged as optional. Given that the "snap geometries to layer" behaves the same way, could the "reference layer" option be set as optional , hence empty by default?

Agreed that would be more intuitive - let's address that in a separate PR...
